coderd/database/queries/insights.sql

```diff
@@ -771,3 +771,124 @@ SELECT
+	SELECT DISTINCT deleted_at AS date
+	FROM user_deleted
+	WHERE deleted_at > @start_time::timestamptz
+		AND deleted_at < @end_time::timestamptz
+
+	UNION
+
+	SELECT @end_time::timestamptz AS date
+),
+-- latest_status_before_range defines the status of each user before the start_time.
+-- We do not include users who were deleted before the start_time. We use this to ensure that
+-- we correctly count users prior to the start_time for a complete graph.
+latest_status_before_range AS (
+    SELECT
+        DISTINCT usc.user_id,
+        usc.new_status,
+        usc.changed_at
+    FROM user_status_changes usc
+	LEFT JOIN user_deleted ud ON usc.user_id = ud.user_id
+    WHERE usc.changed_at < @start_time::timestamptz
+	AND (ud.user_id IS NULL OR ud.deleted_at > @start_time::timestamptz)
+    ORDER BY usc.user_id, usc.changed_at DESC
+),
+-- status_changes_during_range defines the status of each user during the start_time and end_time.
+-- If a user is deleted during the time range, we count status changes prior to the deletion.
+-- Theoretically, it should probably not be possible to update the status of a deleted user, but we
+-- need to ensure that this is enforced, so that a change in business logic later does not break this graph.
+status_changes_during_range AS (
+    SELECT
+        usc.user_id,
+        usc.new_status,
+        usc.changed_at
+    FROM user_status_changes usc
+	LEFT JOIN user_deleted ud ON usc.user_id = ud.user_id
+    WHERE usc.changed_at >= @start_time::timestamptz
+        AND usc.changed_at <= @end_time::timestamptz
+		AND (ud.user_id IS NULL OR usc.changed_at < ud.deleted_at)
+),
+-- relevant_status_changes defines the status of each user at any point in time.
+-- It includes the status of each user before the start_time, and the status of each user during the start_time and end_time.
+relevant_status_changes AS (
+    SELECT
+        user_id,
+        new_status,
+        changed_at
+    FROM latest_status_before_range
+
+    UNION ALL
+
+    SELECT
+        user_id,
+        new_status,
+        changed_at
+    FROM status_changes_during_range
+),
+-- statuses defines all the distinct statuses that were present just before and during the time range.
+-- This is used to ensure that we have a series for every relevant status.
+statuses AS (
+	SELECT DISTINCT new_status FROM relevant_status_changes
+),
+-- We only want to count the latest status change for each user on each date and then filter them by the relevant status.
+-- We use the row_number function to ensure that we only count the latest status change for each user on each date.
+-- We then filter the status changes by the relevant status in the final select statement below.
+ranked_status_change_per_user_per_date AS (
+	SELECT
+	d.date,
+	rsc1.user_id,
+	ROW_NUMBER() OVER (PARTITION BY d.date, rsc1.user_id ORDER BY rsc1.changed_at DESC) AS rn,
+	rsc1.new_status
+	FROM dates_of_interest d
+	LEFT JOIN relevant_status_changes rsc1 ON rsc1.changed_at <= d.date
+)
```

Comment:
I'm a bit worried about the number of CTEs and performance, but I think this is fine for now and let's not prematurely optimize. Just raising this so you're aware that in some cases a CTE performs worse than a pure query with joins and subqueries. That's mainly because the result of a CTE lacks indexes.

Reply:
I was wondering about performance too, but user count and user status changes is going to be way less data than say workspace builds. Like on the order of 1000s. :shrug: 

Might be worth a benchmark, but we don't have a good way to populate "large" datasets atm.

Reply:
I did it this way for legibility. Trying to write the optimal solution here made my own query inscrutable even to myself. I don't consider this to be on a particularly hot path. We do have metrics for this query. If it needs to be optimised, we can do so.

Reply:
> Just raising this so you're aware that in some cases a CTE performs worse than a pure query with joins and subqueries

Also these CTEs were designed to reduce the amount of data being handled as early as possible. We use the existing indices while we have them to identify exactly the data we need ASAP and then once we need to join it all the hope is that we've brought down the volume low enough that it's performant regardless of the lack of indices on CTEs.